### ISSUE TITLE
tomatoware-mipsel compiler unbound Error

### ISSUE THREAD

**Issue by houzi-** [lancethepants/tomatoware#26]
Hello! You do not know what is a good solution 
Error:
```
collect2: error: ld returned 1 exit status
Makefile:318: recipe for target 'unbound-anchor' failed
make: *** [unbound-anchor] Error 1
```
```
 LDFLAGS="-L/mmc/lib -Wl,--gc-sections" CXXFLA
GS="-O3 -mtune=mips32 -mips32 -ffunction-sections -fdata-sections" ./configure -
-prefix=/mmc/unbound --sysconfdir=/mmc/unbound/etc --with-ssl=/mmc --with-libexp
at=/mmc --enable-static-exe --enable-static --disable-shared --disable-flto

make LIBS="-all-static -lcrypto -lz -ldl"
```

**Comment by houzi-** [comment]
arm statically compiled also encountered the same error

**Comment by rampageX** [comment]
compile all libs with: 

`./configure --prefix=/mmc --enable-static --disable-shared`

then unbound:

`LDFLAGS="-Wl,-static -static -static-libgcc -s" ./configure --prefix=/opt --with-ssl=/mmc -with-libexpat=/mmc --disable-flto`

got:

    file /opt/sbin/unbound
    /opt/sbin/unbound: ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), statically linked, stripped

**Comment by houzi-** [comment]
@rampageX  
Thanks! I try

**Comment by houzi-** [comment]
@rampageX 
I do not know why the error occurred , OpenSSL version 1.0.2g
error:
```
configure: error: OpenSSL found in /mmc, but version 0.9.7 or higher is required
```

**Comment by rampageX** [comment]
It seem like tomatoware 1.2 bug, you need recompile and install OpenSSL by yourself

**Comment by houzi-** [comment]
@rampageX 
Thank you! I do not know you compile OpenSSL is the latest version of it?
I was re-statically compiled OpenSSL, but it is also the error! Remove LDFLAGS can be compiled pass

**Comment by houzi-** [comment]
But after` LDFLAGS` can not remove static compiler

**Comment by rampageX** [comment]
download and unpack `openssl-1.0.2g.tar.gz` then:

    ./config --prefix=/mmc shared
    make && make install

**Comment by rampageX** [comment]
when you `make` you may get error:

`/mmc/lib/gcc/arm-buildroot-linux-uclibcgnueabi/5.2.0/../../../../arm-buildroot-linux-uclibcgnueabi/bin/ld: attempted static link of dynamic object `/mmc/lib/libexpat.so'
collect2: error: ld returned 1 exit status`

run:
    mv /mmc/lib/libexpat.la /mmc/lib/libexpat.la.bak
    mv /mmc/lib/libexpat.so /mmc/lib/libexpat.so.bak

continue `make`, it will done.

**Comment by houzi-** [comment]
@rampageX 
Thank you 
It has the perfect solution . OpenSSL is the causes !

**Comment by lancethepants** [comment]
OpenSSL in Tomatoware is built with zlib as a dependency for compression.  If you look at the config.log file when it says it couldn't find OpenSSL, you will see that it fails because it can't find the zlib symbols.  You'd have to ad '-lz' somewhere to help it find it.  For some reason unbound's autoconf is really badly behaved and just fights you the whole way.

**Comment by houzi-** [comment]
@rampageX Thank you! Now I know! Haha！

**Comment by houzi-** [comment]
@lancethepants 
But it is prompted OpenSSL version, rather than not find OpenSSL , OpenSSL version is already higher than this version tips

**Comment by houzi-** [comment]
@lancethepants 
```
LDFLAGS="-Wl,-static -static -static-li
bgcc -lz -s" ./configure --prefix=/mmc/unbound-static --with-ssl=/mmc -with-libe
xpat=/mmc --enable-static-exe --enable-static --disable-shared --disable-flto &&
 make LIBS="-all-static -lcrypto -ldl"
```

Now tomatoware-mipsel statically compiled when the error occurs , I have joined -lz but still can not automatically find

error:
```
checking for HMAC_CTX_init in -lcrypto... no
checking if -lcrypto needs -lgdi32... no
checking if -lcrypto needs -ldl... no
configure: error: OpenSSL found in /mmc, but version 0.9.7 or higher is required
```

**Comment by lancethepants** [comment]
I try to like to find solutions that can be solved with the command line.  In this case we need the linker flags `-lcrytpo -lz` in that order.  Looking at the configure file, I see that this is impossible at the cli, because anything you put at the command line, LDFLAGS or LIBS, will be placed before `-lcrytpo`, so that doesn't work.

What I did to get this working was modify `configure`, replacing all occurences of `-lcrypto` with `lcrypto -lz -ldl`. I also did the same with `-lssl`, but I don't think it's necessary. I used vim's built in find/replace, or sed would be a good way too (or by hand).

You'll hit the libexpat issue as mentioned above, but you can make that work. It does compile unbound, so the libexpat part is something else.

Now whether or not this will segfault I don't know, please let me know.  

Note: These are not issues with tomatoware.  Most projects do not take static compilation into consideration, so it will not always be a straight-forward solution. Most distros use glibc which does not even allow for static compilation.  Sometimes patching has been needed for static compilation to work.  lighttpd is a famous example.  Here's another example of a cgit patch I needed for static linking to work and for the binary to function properly.

https://lists.zx2c4.com/pipermail/cgit/2015-January/002328.html

**Comment by houzi-** [comment]
@lancethepants 
OpenSSL solve the problem , but there have been statically compiled `expat` library file issue
```
/mmc/lib/gcc/mipsel-buildroot-linux-uclibc/5.2.0/../../../../mipsel-buildroot-linux-uclibc/bin/ld: attempted static link of dynamic object `/mmc/lib/libexpat.so'
collect2: error: ld returned 1 exit status
Makefile:303: recipe for target 'unbound' failed
make: *** [unbound] Error 1
```

**Comment by lancethepants** [comment]
rampageX has a solution for this up above.  This is really a unbound autoconf issue.

**Comment by houzi-** [comment]
But this can not be resolved, or an error

**Comment by houzi-** [comment]
@lancethepants
 1.5.9 version of this treatment can be compiled pass
1.5.8 seemingly not pass

**Comment by lancethepants** [comment]
Well I released Tomatoware v1.4 with gcc-6.1 you can try out. I successfully compile dozens of static binaries that work fine, so I'm still of the opinion that this is a unbound issue at the core. Have you tried asking them about building static binary?
